buildSrc/src/main/java/gradlebuild/JniNokeePlugin.java

```diff
@@ -175,6 +225,158 @@ private void configureNativeVersionGeneration(Project project) {
         );
     }
 
+    private void configureNativeJars(Project project, VariantsExtension variants, boolean testVersionFromLocalRepository) {
+        TaskProvider<Jar> emptyZip = project.getTasks().register("emptyZip", Jar.class, jar -> jar.getArchiveClassifier().set("empty"));
+        // We register the publications here, so they are available when the project is used as a composite build.
+        final Provider<String> artifactId = project.getTasks().named("jar", Jar.class)
+            .flatMap(Jar::getArchiveBaseName);
+        final Provider<Iterable<Jar>> publishableJarTasks = library(project).flatMap(allVariants())
+            .map(toPublishableVariantGroups())
+            .map(toBuildableVariantUsingVariantNames(variants.getVariantNames()))
+            .flatMap(toPublishableJars(project, artifactId));
+
+        project.getConfigurations().named("runtimeElements", configuration -> {
+            final Provider<? extends Iterable<PublishArtifact>> publishableJars = project.getObjects().listProperty(PublishArtifact.class).value(publishableJarTasks.map(it -> stream(it).map(ArchivePublishArtifact::new).collect(toList())));
+            configuration.getOutgoing().getArtifacts().addAllLater(publishableJars);
+        });
+
+        project.afterEvaluate(ignored -> {
+            for (Jar nativeJar : publishableJarTasks.get()) {
+                final String variantName = nativeJar.getName().substring("jar-".length());
+                project.getExtensions().configure(PublishingExtension.class, publishing -> publishing.publications(publications -> publications.create(variantName, MavenPublication.class, publication -> {
+                    publication.artifact(nativeJar);
+                    publication.artifact(emptyZip.get(), it -> it.setClassifier("sources"));
+                    publication.artifact(emptyZip.get(), it -> it.setClassifier("javadoc"));
+                    publication.setArtifactId(nativeJar.getArchiveBaseName().get());
+                })));
+            }
+        });
```

Comment:
🔍 We could eliminate the `afterEvaluate` if we could successfully use `addAllLater` and map our `publishableTasks` provider into a list of publications. The alternative that we experimented with in Nokee is a lazy stream-like interface that allows to either collect into a `Provider` or iterate as new values come into existence. It would be very similar and would look something like the following:

```
Stream<Jar> publishableJarTaks = ...
publishableJarTasks.forEach { nativeJar ->
    project.publishing.publications.create('<name>', MavenPublication) {
        artifact(nativeJar)
        //...
    }
}
```

Reply:
The streaming concept is something we are looking into adding to Nokee. We unfortunately don't have an ETA.